Support parameter substitution for all string fields in Template

mfojtik opened the pull request:
This patch allow generic substitution for **all** `string` fields in any object we have. It allows referencing parameters in fields like 'host', 'output->registry', annotations, etc...

The current state is that we support substitution only for Env fields and only for certain types. There is also no documentation that would say this... I think it would be better to say:

"we support parameter substitution for **all** string fields" vs "we support parameter substitution for all Env fields defined for these types: ...."

mfojtik commented:
@smarterclayton @bparees fyi

this was discussed many times and it pops up everytime we need to hardcode IP or hostname into items... I tried to get familiar with reflect during weekend, and I use this as a good exercise...

bparees commented on pkg/util/struct.go:
```diff
@@ -0,0 +1,48 @@
+
+// VisitAllStringFields visits all string fields in the given pointer to struct
+// and allow modifying them with visitor function
+func VisitAllStringFields(obj interface{}, visitor func(string) string) {
```
why not make this take a type argument so it can generically visit any field type?

and why not have the visitor take the val object and do what it wants with it, rather than having the visitor return a value to be set?   I don't think this is a true visitor pattern.

mfojtik replied:
i tried to make this specific for 'string' values. If I make this more generic, you will have to protect yourself using CanSet() etc and it will make the visitor() func more complex.

soltysh commented on pkg/template/helper.go:
```diff
@@ -0,0 +1,48 @@
+			if val.CanInterface() {
+				VisitAllStringFields(val.Addr().Interface(), visitor)
+			}
+		default:
```
I don't see any point to adding this here, since you're not using it.

mfojtik replied:
it is just a placeholder for comment ;-) but I can remove it

soltysh commented:
Generally speaking LGTM :+1:

smarterclayton commented:
I thought we agreed we are not doing this.  Why are we doing this?

bparees commented:
@smarterclayton @mfojtik got bored.  I warned him you would not approve. :)

that said, this actually makes our substitution logic cleaner and more consistent, so I like the change, even if we don't want to advertise the capability.  it's still not the full templating language you fear us creating.

mfojtik commented:
@smarterclayton we don't have to, I had the code for this already so I
though I will not keep it for myself...

and there are plenty use cases for this already.
On Feb 2, 2015 6:51 PM, "Clayton Coleman" <notifications@github.com> wrote:

> I thought we agreed we are not doing this. Why are we doing this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/827#issuecomment-72502813>.
>

smarterclayton commented:
If you have this templating, you are forcing people who build a template to know that escaping needs to be done on arbitrary fields.  That complicates tools that build templates, which app-gen and new-app will be.  

Let's leave this on the side for now, and we'll come back to it when we have a use case that can only be addressed by this septa ion.


> On Feb 2, 2015, at 12:56 PM, Ben Parees <notifications@github.com> wrote:
> 
> @smarterclayton @mfojtik got bored. I warned him you would not approve. :)
> 
> that said, this actually makes our substitution logic cleaner and more consistent, so I like the change, even if we don't want to advertise the capability. it's still not the full templating language you fear us creating.
> 
> —
> Reply to this email directly or view it on GitHub.
>

mfojtik commented:
if we introduce this later, aren't we going to break the tools then? if we
do it now, template tools writers will know that they need to escape now.
On Feb 2, 2015 7:16 PM, "Clayton Coleman" <notifications@github.com> wrote:

> If you have this templating, you are forcing people who build a template
> to know that escaping needs to be done on arbitrary fields. That
> complicates tools that build templates, which app-gen and new-app will be.
>
> Let's leave this on the side for now, and we'll come back to it when we
> have a use case that can only be addressed by this septa ion.
>
>
> > On Feb 2, 2015, at 12:56 PM, Ben Parees <notifications@github.com>
> wrote:
> >
> > @smarterclayton @mfojtik got bored. I warned him you would not approve.
> :)
> >
> > that said, this actually makes our substitution logic cleaner and more
> consistent, so I like the change, even if we don't want to advertise the
> capability. it's still not the full templating language you fear us
> creating.
> >
> > —
> > Reply to this email directly or view it on GitHub.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/827#issuecomment-72507714>.
>

smarterclayton commented:
If there was a clear use case *requiring* this functionality now, I'd agree.  However, we haven't even defined that.  But at least one way is to make the "replace everything" not be the default behavior, but be explicit (replace environment, object references, selectors, etc).

----- Original Message -----
> if we introduce this later, aren't we going to break the tools then? if we
> do it now, template tools writers will know that they need to escape now.
> On Feb 2, 2015 7:16 PM, "Clayton Coleman" <notifications@github.com> wrote:
> 
> > If you have this templating, you are forcing people who build a template
> > to know that escaping needs to be done on arbitrary fields. That
> > complicates tools that build templates, which app-gen and new-app will be.
> >
> > Let's leave this on the side for now, and we'll come back to it when we
> > have a use case that can only be addressed by this septa ion.
> >
> >
> > > On Feb 2, 2015, at 12:56 PM, Ben Parees <notifications@github.com>
> > wrote:
> > >
> > > @smarterclayton @mfojtik got bored. I warned him you would not approve.
> > :)
> > >
> > > that said, this actually makes our substitution logic cleaner and more
> > consistent, so I like the change, even if we don't want to advertise the
> > capability. it's still not the full templating language you fear us
> > creating.
> > >
> > > —
> > > Reply to this email directly or view it on GitHub.
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/openshift/origin/pull/827#issuecomment-72507714>.
> >
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/827#issuecomment-72510985

mfojtik commented:
@smarterclayton the use-case I was thinking about is to parametrize the image name (with the registry IP address, that is currently hard coded). However, I know that we are going to solve this with allowing to configure the lookup order for the images, so you can specify the list of registries...

Another use-case is the Route (see the test-case I updated in this example). I think we would need to parametrize the DNS name (and we would most likely have special generator for that in future)...

I agree that for time being, there is no use-case that "require" this functionality. This PR is made in a way that we can merge this anytime we need (does not need heavy rebasing). So we can re-open this when the time come ;-)

mfojtik commented:
@bparees @smarterclayton I re-opened this pull request and rebased it to latest master. I also fixed tests and the code (check the guestbook.json example, which is a bit extreme but it provides good coverage for the replacement in various places).

@kargakis PTAL as well as you were working on CLI/templating stuff recently.

kargakis commented on pkg/template/template.go:
```diff
@@ -85,51 +83,25 @@ func GetParameterByName(t *api.Template, name string) *api.Parameter {
-// TODO: Implement substitution for more types and fields.
 func (p *Processor) SubstituteParameters(params []api.Parameter, item runtime.Object) (runtime.Object, error) {
-	// Make searching for given parameter name/value more effective
+		// Make searching for given parameter name/value more effective
```
Is this a vim thing? :)

kargakis commented:
LGTM

smarterclayton commented on pkg/template/helper.go:
```diff
@@ -0,0 +1,48 @@
+			if val.Elem().Kind() == reflect.Struct {
+				VisitFields(val.Interface(), visitor)
+			}
+		case reflect.Struct:
```
Numbers?

smarterclayton replied:
Also need to handle arrays and nil pointers (i don't know if you have a test case for that).

smarterclayton replied:
You can have an array of arrays, and an array of maps, so this code doesn't look right to me.

mfojtik replied:
numbers are tricky, can you have parameter reference in json in a number
field without quotes?
On Apr 17, 2015 7:59 PM, "Clayton Coleman" <notifications@github.com> wrote:

> In pkg/template/helper.go
> <https://github.com/openshift/origin/pull/827#discussion_r28615837>:
>
> > +// visitor function on it.
> > +// The visitor function can manipulate the fields values.
> > +func VisitFields(obj interface{}, visitor func(string) string) {
> > +	v := reflect.ValueOf(obj).Elem()
> > +
> > +	visit := func(val reflect.Value) {
> > +		switch val.Kind() {
> > +		case reflect.String:
> > +			if val.CanSet() {
> > +				val.SetString(visitor(val.String()))
> > +			}
> > +		case reflect.Ptr, reflect.Interface:
> > +			if val.Elem().Kind() == reflect.Struct {
> > +				VisitFields(val.Interface(), visitor)
> > +			}
> > +		case reflect.Struct:
>
> Numbers?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/827/files#r28615837>.
>

mfojtik replied:
@smarterclayton also nil pointers should be handled (there are several in the guestbook.json template). I can make explicit nil unit test case. Arrays of arrays/maps needs to be added.

kargakis replied:
Maps of maps are also quite common.

kargakis replied:
Even though I am not sure we can have those in templates.

smarterclayton replied:
Maybe not - if there's not a clean solution we can punt on it for now.

> On Apr 17, 2015, at 2:18 PM, Michal Fojtik <notifications@github.com> wrote:
> 
> In pkg/template/helper.go:
> 
> > +// visitor function on it.
> > +// The visitor function can manipulate the fields values.
> > +func VisitFields(obj interface{}, visitor func(string) string) {
> > +	v := reflect.ValueOf(obj).Elem()
> > +
> > +	visit := func(val reflect.Value) {
> > +		switch val.Kind() {
> > +		case reflect.String:
> > +			if val.CanSet() {
> > +				val.SetString(visitor(val.String()))
> > +			}
> > +		case reflect.Ptr, reflect.Interface:
> > +			if val.Elem().Kind() == reflect.Struct {
> > +				VisitFields(val.Interface(), visitor)
> > +			}
> > +		case reflect.Struct:
> numbers are tricky, can you have parameter reference in json in a number field without quotes?
> …
> —
> Reply to this email directly or view it on GitHub.
>

smarterclayton replied:
We should probably add the test now and fix it so it doesn't break.

> On Apr 17, 2015, at 4:17 PM, Michail Kargakis <notifications@github.com> wrote:
> 
> In pkg/template/helper.go:
> 
> > +// visitor function on it.
> > +// The visitor function can manipulate the fields values.
> > +func VisitFields(obj interface{}, visitor func(string) string) {
> > +	v := reflect.ValueOf(obj).Elem()
> > +
> > +	visit := func(val reflect.Value) {
> > +		switch val.Kind() {
> > +		case reflect.String:
> > +			if val.CanSet() {
> > +				val.SetString(visitor(val.String()))
> > +			}
> > +		case reflect.Ptr, reflect.Interface:
> > +			if val.Elem().Kind() == reflect.Struct {
> > +				VisitFields(val.Interface(), visitor)
> > +			}
> > +		case reflect.Struct:
> Even though I am not sure we can have those in templates.
> 
> —
> Reply to this email directly or view it on GitHub.
>

mfojtik commented:
@kargakis @smarterclayton I added support for array-in-array, map-in-map and moved everything into `pkg/util` so it can be re-used. I also updated unit tests to cover the cases above. The nil Ptr is already covered.

mfojtik commented:
[test]

mfojtik commented:
@smarterclayton @kargakis one more change, I changed the name to 'VisitObjectStrings' as this func can also substitute arrays/etc (not just structs). Also I switched to use table tests (which looks pretty cool now). Please let me know if you want to add more test cases.

kargakis commented on pkg/util/object.go:
```diff
@@ -0,0 +1,48 @@
+		}
+	}
+
+	switch v.Kind() {
```
Shouldn't you add a case for Arrays and Slices here as well and do the indexing (as you already do with Maps) instead of doing it in the visitor function? IMO indexing of any struct, built-in or not, should happen in the same place, either the visitor function or here. Actually it should probably happen here...

kargakis commented on pkg/util/object_test.go:
```diff
@@ -0,0 +1,61 @@
+	Array        []string
+}
+
+func TestVisitObjectStringsOnStruct(t *testing.T) {
```
[Nit] Name of the test should be `Test+(Name of the tested function)`

mfojtik replied:
@kargakis I was planning to add more test functions (OnStruct, onArray, onMap)....

mfojtik commented:
@kargakis fixed, thanks for comments :-)

smarterclayton commented:
I'm one with not supporting integers for now.

> On Apr 20, 2015, at 5:11 AM, Michal Fojtik <notifications@github.com> wrote:
> 
> @kargakis @smarterclayton I added support for array-in-array, map-in-map and moved everything into pkg/util so it can be re-used. I also updated unit tests to cover the cases above. The nil Ptr is already covered.
> 
> —
> Reply to this email directly or view it on GitHub.
>

mfojtik commented:
@smarterclayton +1

We can support that later by specifying the JSON path in the Parameter definition (iow. "this is a number that should be places here <>")

smarterclayton commented on pkg/util/object.go:
```diff
@@ -0,0 +1,53 @@
+package util
```
Put this in its own util package util/jsonreplace or something

mfojtik replied:
done.

smarterclayton commented on pkg/template/template.go:
```diff
@@ -5,16 +5,15 @@ import (
 	"github.com/openshift/origin/pkg/template/api"
 	. "github.com/openshift/origin/pkg/template/generator"
 	"github.com/openshift/origin/pkg/util"
+	"github.com/openshift/origin/pkg/util/stringreplace"
```
objectreplace

bparees commented on pkg/util/stringreplace/object.go:
```diff
@@ -0,0 +1,53 @@
+			visitField(v.Field(i))
+		}
+	case reflect.Map:
+		for _, k := range v.MapKeys() {
```
not certain there's a strong use case, but what about substituting keys in maps, not just values?

mfojtik replied:
I think Clayton was against doing it in keys and I think I sever.
On Apr 20, 2015 5:11 PM, "Ben Parees" <notifications@github.com> wrote:

> In pkg/util/stringreplace/object.go
> <https://github.com/openshift/origin/pull/827#discussion_r28696961>:
>
> > +			}
> > +			val.SetString(visitor(val.String()))
> > +		}
> > +	}
> > +
> > +	switch v.Kind() {
> > +	case reflect.Slice, reflect.Array:
> > +		for c := 0; c < v.Len(); c++ {
> > +			visitField(v.Index(c))
> > +		}
> > +	case reflect.Struct:
> > +		for i := 0; i < v.NumField(); i++ {
> > +			visitField(v.Field(i))
> > +		}
> > +	case reflect.Map:
> > +		for _, k := range v.MapKeys() {
>
> not certain there's a strong use case, but what about substituting keys in
> maps, not just values?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/827/files#r28696961>.
>

mfojtik replied:
agree..  android autocomplete.. .
On Apr 20, 2015 5:17 PM, "Michal Fojtik" <mi@mifo.sk> wrote:

> I think Clayton was against doing it in keys and I think I sever.
> On Apr 20, 2015 5:11 PM, "Ben Parees" <notifications@github.com> wrote:
>
>> In pkg/util/stringreplace/object.go
>> <https://github.com/openshift/origin/pull/827#discussion_r28696961>:
>>
>> > +			}
>> > +			val.SetString(visitor(val.String()))
>> > +		}
>> > +	}
>> > +
>> > +	switch v.Kind() {
>> > +	case reflect.Slice, reflect.Array:
>> > +		for c := 0; c < v.Len(); c++ {
>> > +			visitField(v.Index(c))
>> > +		}
>> > +	case reflect.Struct:
>> > +		for i := 0; i < v.NumField(); i++ {
>> > +			visitField(v.Field(i))
>> > +		}
>> > +	case reflect.Map:
>> > +		for _, k := range v.MapKeys() {
>>
>> not certain there's a strong use case, but what about substituting keys
>> in maps, not just values?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/openshift/origin/pull/827/files#r28696961>.
>>
>

smarterclayton replied:
Yeah, I don't want us to be meta programming for now.  We can always discuss later with a concrete use case.  Templates are for precanned apps, not for on the fly generation.  The API should not be using keys for values we need to change (should be using arrays instead).

> On Apr 20, 2015, at 11:17 AM, Michal Fojtik <notifications@github.com> wrote:
> 
> In pkg/util/stringreplace/object.go:
> 
> > +			}
> > +			val.SetString(visitor(val.String()))
> > +		}
> > +	}
> > +
> > +	switch v.Kind() {
> > +	case reflect.Slice, reflect.Array:
> > +		for c := 0; c < v.Len(); c++ {
> > +			visitField(v.Index(c))
> > +		}
> > +	case reflect.Struct:
> > +		for i := 0; i < v.NumField(); i++ {
> > +			visitField(v.Field(i))
> > +		}
> > +	case reflect.Map:
> > +		for _, k := range v.MapKeys() {
> I think Clayton was against doing it in keys and I think I sever.
> …
> —
> Reply to this email directly or view it on GitHub.
>

bparees commented:
lgtm.

smarterclayton commented:
[merge]

bparees commented:
fixes https://github.com/openshift/origin/issues/1636

smarterclayton commented:
This did not provide escaping, so there is no way for an author to provide a string that is evaluated at run time (vs earlier).  I think provisionally I'm going to handle it with `$${` being escaped like printf.
